src/main/java/io/aiven/kafka/auth/audit/Auditor.java

```diff
@@ -0,0 +1,152 @@
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class Auditor implements Configurable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Auditor.class);
+
+    protected Map<AuditKey, UserActivity> auditStorage = new HashMap<>();
+
+    private final Lock auditLock = new ReentrantLock();
+
+    private final ScheduledExecutorService auditScheduler =
+        Executors.newScheduledThreadPool(1);
+
+    public Auditor() {
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        final AuditorConfig auditorConfig = new AuditorConfig(configs);
+        auditScheduler.scheduleAtFixedRate(
+            this::dump,
+            auditorConfig.getAggregationPeriodInSeconds(),
+            auditorConfig.getAggregationPeriodInSeconds(),
+            TimeUnit.SECONDS
+        );
+    }
+
+    public void addActivity(final Session session,
+                            final Operation operation,
+                            final Resource resource,
+                            final Boolean hasAccess) {
+
+        final AuditKey auditKey = new AuditKey(session.principal(), session.clientAddress());
+        auditLock.lock();
+        try {
+            auditStorage.compute(auditKey, (key, userActivity) -> {
+                if (Objects.isNull(userActivity)) {
+                    return onUserActivity(new UserActivity(), operation, resource, hasAccess);
+                } else {
+                    return onUserActivity(userActivity, operation, resource, hasAccess);
+                }
+            });
+        } finally {
+            auditLock.unlock();
+        }
+    }
+
+    public void stop() {
+        auditScheduler.shutdownNow();
+        dump();
+    }
+
+    protected void dump() {
+        final Map<AuditKey, UserActivity> auditStorageDump;
+        auditLock.lock();
+        try {
+            auditStorageDump = auditStorage;
+            auditStorage = new HashMap<>();
+        } finally {
+            auditLock.unlock();
+        }
+        auditStorageDump.forEach((key, userActivity) -> LOGGER.info(auditMessage(key, userActivity)));
+    }
+
+    protected String auditMessage(final AuditKey key, final UserActivity userActivity) {
+        final StringBuilder auditMessage = new StringBuilder(key.principal.toString());
+        auditMessage
+            .append(" (").append(key.sourceIp).append(")")
+            .append(" was active since ")
+            .append(userActivity.activeSince.format(DateTimeFormatter.ISO_INSTANT));
+        if (userActivity.hasOperations()) {
+            auditMessage.append(": ")
+                .append(
+                    userActivity
+                        .operations
+                        .stream()
+                        .map(this::userOperationMessage)
+                        .collect(Collectors.joining(", "))
+                ).append(".");
```

Comment:
This `.` appear only when there are operations. Personally, I'd prefer to remove it at all in this case

Reply:
true